inception-imls-opennlp/src/main/java/de/tudarmstadt/ukp/inception/recommendation/imls/opennlp/doccat/OpenNlpDoccatRecommender.java

```diff
@@ -89,10 +90,18 @@ public void train(RecommenderContext aContext, List<CAS> aCasses)
         params.put(BeamSearch.BEAM_SIZE_PARAMETER, Integer.toString(beamSize));
         
         DoccatModel model = train(nameSamples, params);
-        if (model != null) {
-            aContext.put(KEY_MODEL, model);
-            aContext.markAsReadyForPrediction();
+        
+        if (model == null) {
```

Comment:
Also here, will this recommender be able to predict if it is not trained?

Reply:
> If a recommender needs to be trained (TRAINING_REQUIRED), but no model is available, I think throwing an exception makes sense. If a recommender works also without training (TRAINING_SUPPORTED), then failure to train a model with new data does not lead to an exception.

At prediction time yes - throwing an exception would make sense. But at training time, if there is not sufficient data to train a model, then I think throwing an exception overshoots a bit.

I have prepared an update for this PR with the following changes:

* RecommenderContext is now an AutoClosable and will *always* be closed by TrainingTask now
* the decision whether a context is ready for training is now made by the RecommendationEngine at prediction time in accordance with the training capabilities it declares
* no exceptions thrown if a model could not be trained unless there was an actual error during training
* isTrainable means TRAINING_SUPPORTED not TRAINING_REQUIRED

Reply:
Also, my update locks down the RecommenderContext once it is closed so that now data can be added anymore. This presently causes exceptions to be thrown because of #1276.

Reply:
Wow, ok. This looks good (and I learned something new)! This feature exploded a bit from my first commit...